src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.Encrypt.cs

```diff
@@ -266,36 +251,54 @@ private static CMSG_RECIPIENT_ENCODE_INFO EncodeRecipientInfo(CmsRecipient recip
 
                     pEncodeInfo->cbSize = sizeof(CMSG_KEY_TRANS_RECIPIENT_ENCODE_INFO);
 
-                    if (recipient.RSAEncryptionPadding is null)
+                    RSAEncryptionPadding padding = recipient.RSAEncryptionPadding;
+
+                    if (padding is null)
                     {
-                        CRYPT_ALGORITHM_IDENTIFIER algId = pCertInfo->SubjectPublicKeyInfo.Algorithm;
-                        pEncodeInfo->KeyEncryptionAlgorithm = algId;
+                        if (recipient.Certificate.GetKeyAlgorithm() == Oids.RsaOaep)
+                        {
+                            byte[] parameters = recipient.Certificate.GetKeyAlgorithmParameters();
+
+                            if (parameters == null || parameters.Length == 0)
+                            {
+                                padding = RSAEncryptionPadding.OaepSHA1;
+                            }
+                            else if (!PkcsHelpers.TryGetRsaOaepEncryptionPadding(parameters, out padding, out _))
```

Comment:
It didn't seem correct to me to blindly shove the certificate's parameters in to `pEncodeInfo->KeyEncryptionAlgorithm.Parameters` without decoding it to make sure it was something sensible. If my concern doesn't seem warranted, we can do that instead.